### PR TITLE
Add pure virtual function name() in AbstractRule

### PR DESCRIPTION
**Pull request by julianmenzler**
This PR adds `AbstractRule::name`, similar to `AbstractOperator::name`. 

It replaces 

```cpp
auto& rule_reference = *rule;
std::string(typeid(rule_reference).name());
```

with

```cpp
rule->name();
```

### CODE REVIEW

File: src/lib/optimizer/strategy/between_composition_rule.cpp
Comment (Bouncner) [inline src/lib/optimizer/strategy/between_composition_rule.cpp:40]: Would it make sense to use proper names with spaces here?

**Comment by Bensk1** [comment]
How is this done for our operators? Do we want to have it analogously?

Am 29.06.2021 um 07:45 schrieb Bouncner ***@***.***>:

﻿

@Bouncner commented on this pull request.

________________________________

In src/lib/optimizer/strategy/between_composition_rule.cpp<https://github.com/hyrise/hyrise/pull/2378#discussion_r660297681>:

> @@ -36,6 +36,11 @@ PredicateCondition get_between_predicate_condition(bool left_inclusive, bool rig

 namespace opossum {

+std::string BetweenCompositionRule::name() const {
+  static const auto name = std::string{"BetweenCompositionRule"};


Would it make sense to use proper names with spaces here?

File: src/lib/optimizer/strategy/between_composition_rule.cpp
Comment (julianmenzler) [reply src/lib/optimizer/strategy/between_composition_rule.cpp:40]: > Would it make sense to use proper names with spaces here?

I do not see any benefits.

> How is this done for our operators? Do we want to have it analogously?

Yes. `name()` outputs the operator's class name in PascalCase.

Comment (Bensk1) [reply src/lib/optimizer/strategy/between_composition_rule.cpp:40]: > I do not see any benefits.

Me neither.

Comment (Bouncner) [reply src/lib/optimizer/strategy/between_composition_rule.cpp:40]: Well, it's the way the english language works. :flushed:

**Review by Bouncner — APPROVED** [review]
Looks good to me. Since @Bensk1 also looked at the PR, please wait for his approval as well.

File: src/test/lib/optimizer/optimizer_test.cpp
Comment (Bensk1) [inline src/test/lib/optimizer/optimizer_test.cpp:228]: Why has this changed? If it stays like this, should it become `_counter`?

Comment (julianmenzler) [reply src/test/lib/optimizer/optimizer_test.cpp:226]: Reverted by now.

**Review by Bensk1 — APPROVED** [review]

**Comment by Bensk1** [comment]
Ready for merge?!

**Pull request merged**
